client/main.go

```diff
@@ -24,6 +24,47 @@ var (
 	EventsClient   events.EventsServiceClient
 )
 
+// MetaClient aggregates all the grpc clients available from Tinkerbell Server
+type MetaClient struct {
+	TemplateClient template.TemplateServiceClient
+	WorkflowClient workflow.WorkflowServiceClient
+	HardwareClient hardware.HardwareServiceClient
+	EventsClient   events.EventsServiceClient
+}
+
+// NewMetaClientFromGlobal is a dirty hack that returns a MetaClient using the
+// global variables exposed by the client package. Globals should be avoided
+// and we will deprecated them at some point replacing this function with
+// NewMetaClient. If you are strating a new project please use the last one
+func NewMetaClientFromGlobal() (*MetaClient, error) {
+	// This is required because we use init() too often, even more in the
+	// CLI and based on where you are sometime the clients are not initialised
+	if TemplateClient == nil {
+		err := Setup()
+		if err != nil {
+			panic(err)
+		}
+	}
+	return &MetaClient{
+		TemplateClient: TemplateClient,
+		WorkflowClient: WorkflowClient,
+		HardwareClient: HardwareClient,
+		EventsClient:   EventsClient,
+	}, nil
+}
+
+// NewMetaClient returns a MetaClient. A structure that contains all the
+// clients made available from tink-server. This is the function you should use
+// instead of NewMetaClientFromGlobal that will be deprecated soon
+func NewMetaClient(conn grpc.ClientConnInterface) *MetaClient {
+	return &MetaClient{
+		TemplateClient: template.NewTemplateServiceClient(conn),
+		WorkflowClient: workflow.NewWorkflowServiceClient(conn),
+		HardwareClient: hardware.NewHardwareServiceClient(conn),
+		EventsClient:   events.NewEventsServiceClient(conn),
+	}
+}
+
 // GetConnection returns a gRPC client connection
 func GetConnection() (*grpc.ClientConn, error) {
 	certURL := os.Getenv("TINKERBELL_CERT_URL")
```

Comment:
Is it possible to make the client pkg independent of environment variables and hoist them to code using the client pkg code and insert them as arguments in the GetConnection function?? It will make the client pkg more independent of type of use (imo)...

Reply:
It is already like that if you use the function `NewMetaClient` https://github.com/tinkerbell/tink/pull/406/files#diff-39e4af22263d62c01d979fa1f6ec1770d58fc1b03b0aeec2f17ba84e7136c0dbR59 you have to pass a `grpc.ClientConnInterface` to it, the way you make it is totally up to you

Reply:
@gianarb I know, but then the user of the pkg needs to implement the CERT stuff himself and more or less is duplicating 90% of the code in this function.

To be honest, I'm lazy and spoiled by the k8s client pkg api's... ;-)

Reply:
I am even not understanding or I am doing some wrong assumption. Let me try to explain.

We can discuss if we like how GetConneciton works #174 does it. if you want to get a GRPC connection up without an environment variable you CAN do it with this PR. Look at `GetConnection` as a utility function you can use if you want to have a consistent UX with other Tinkerbell project. If you are speaking about something else or it looks mandatory to you can you elaborate? because maybe I am looking at this problem in the wrong direction

Reply:
Ok, didn't want to confuse you! 

>  Look at GetConnection as a utility function you can use if you want to have a consistent UX with other Tinkerbell project. If you are speaking about something else or it looks mandatory to you can you elaborate?

I see GetConnection as a utility function too but I like to have a more generic kind that doesn't force ENV vars but enables the use of the certificate config. If I don't want to use the specific ENV vars (but with the CERT loading part) I need to implement the whole function, including CERT loading etc myself.

> We can better encapsulate the GetConnection to accept a struct that gets populated via env var, that's probably the right solution but I will probably do it later on in its own chunk of work

Yep, thats the ultimate implementation I was looking for: struct argument for a 'GetConnectionWithConfig' function that also uses the CERT stuff. But I do understand if that is maybe implemented later in its own chunk of work.

Thanks for taking the time to answer me, great job!! 👍

Reply:
/minor note: I *really* want to get rid of all the odd cert handling stuff :D.